### PR TITLE
document kubelet --node-ip and external cloud provider node addresses logic

### LINKED ISSUE #56: Kubelet no longer restricts InternalIP to --node-ip after upgrade to CCM
This affects certain common configurations on OpenStack, but I don't think the issue is specific to OpenStack so I am reporting it here.

A worker has 2 attached networks:

* The machine network: all k8s traffic lives here
* A storage network

There is no routing between these 2 networks.

Both the legacy and external CCM return both of these addresses as InternalIP in NodeAddresses (I will discuss OpenStack CCM's internal-network-name below). However, kubelet has different logic for filtering NodeAddresses depending on their source:

**[Legacy cloud provider](https://github.com/kubernetes/kubernetes/blob/b6c06a95d7ff262a876b95a1035eaf478a7cb961/pkg/kubelet/nodestatus/setters.go#L125-L147)**: kubelet filters NodeAddresses before writing them to the Node. It ensures that the IP given in --node-ip is present in the list of NodeAddresses, and then _it removes all other NodeAddresses of that type_.

**[External CCM](https://github.com/kubernetes/kubernetes/blob/b6c06a95d7ff262a876b95a1035eaf478a7cb961/pkg/kubelet/nodestatus/setters.go#L97-L110)**: kubelet annotates the Node with the IP given in --node-ip, but explicitly does not alter NodeAddresses. [CCM will raise a sync error](https://github.com/kubernetes/cloud-provider/blob/b25f66fb82d43c45e81d79c8250a5af0fb128f32/controllers/node/node_controller.go#L354-L357) if this annotation is invalid but it does not filter NodeAddresses based on it.

Assuming `--node-ip` is set correctly on the kubelet command line, the impact is that without additional configuration (I will discuss OpenStack CCM's internal-network-name below), kubelet will use both the machine network and storage network when using an external CCM, but only the machine network when using a legacy cloud provider. Numerous things break when kubelet itself or pods using host networking use an IP which is not generally routable within the cluster (specifically not on the control plane, where the storage network is not attached). Specifically the Node will not become Ready, and any proxy operation (e.g. logs, exec) to the node will fail because apiserver can't route to the published IP.

OpenStack CCM has an `internal-network-names` config directive which would allow us to restrict the NodeAddresses returned by OpenStack CCM specifically to those which match the given network names. I have some issues with the implementation of this directive, but despite this I think it could work for us if we were able to use it. The principal issue with it is that we don't have an existing, unambiguous, canonical source we can use to automatically populate it. 'We' in this context is OpenShift, but even a CAPO provisioned cluster would have the same issue as worker machinesets are not required to be attached to the cluster-provisioned network, they just have to be able to route to it. This would be the expected configuration in a spine-leaf deployment, for example. So the only context we have for certain is the list of attached networks, which we can get from MachineSet (or Machine). But these define the networks, not their semantics. There is currently no metadata which unambiguously identifies one or more of the networks as being usable for InternalIP. We are left with only heuristics to differentiate the machine network from the storage network, which is not ideal. Incidentally this is a problem which I would also very much like to address, but I think it is a separate problem.

For OpenShift clusters deployed on OpenStack, we currently determine `--node-ip` on-host by looking at local IPs and testing which of them can be used to access the internal VIP. This is also janky if I'm honest, but it's existing, battle-hardened jank that doesn't require new metadata.

The user experience we want here is that the user can upgrade from legacy to external CCM without manual intervention. We can automatically transform cloud.conf during this process, but only if we have the information to do it. Bizarrely, we don't seem to have a canonical list of machine networks anywhere, so we would be left either using heuristics or asking the user to enter them. `--node-ip` is an existing mechanism that achieves this and is already well tested.

Questions:
* What is the contract of `--node-ip`?
* Should we be restricting NodeAddress to `--node-ip` in external CCM the same as it's currently done for legacy cloud provider?
* Should kubelet do this, or the CCM?
* If the CCM should do this, should the behaviour be optional?
* If we explicitly don't want to carry this behaviour of `--node-ip` to the external CCM, is there a better way to fix the described situation that I haven't thought of?

### PR DESCRIPTION
**Pull request by aojea**
/kind bug
/kind api-change

#### What this PR does / why we need it:

  document kubelet node-ip with cloud provider external
    
    The node.status.addresses logic grew organically and with weird
    semantics, this commit try to document existing semantics when
    the kubelet uses an external cloud provider.
    
    The node.status.addresses can be populated by the kubelet at startup or
    delegated to the external cloud provider.
    
    If the --node-ip flag is set to an IP in the node, the kubelet will add
    an annotation to the Node object that will be respected by the external
    cloud providers, no new IP addresses will be added for the same address
    type.
    
    If the IP set in the --node-ip flag is `0.0.0.0` or `::`, the kubelet
    will initialize the node with the default address of the corresponding
    IP family of the unspecified address, and the cloud-provider will override
    it later.

This behavior is inherited from these PRs 
- https://github.com/kubernetes/kubernetes/pull/107750 , fixed https://github.com/kubernetes/cloud-provider/issues/56 that says
> Legacy cloud provider: kubelet filters NodeAddresses before writing them to the Node. It ensures that the IP given in --node-ip is present in the list of NodeAddresses, and then it removes all other NodeAddresses of that type.
- https://github.com/kubernetes/kubernetes/pull/80003

> For example, if nodeIP is 10.0.0.1 and cloud provider returns an InternalIP 10.0.0.1, then other InternalIPs are discarded from enforcedNodeAddresses, as expected. But this bug makes it so that the cloud provider's hostname is overridden too.

- https://github.com/kubernetes/kubernetes/pull/65594
- 
This seems when it originated the filtering behavior, however, but there is need to apply the same legacy behavior to the external cloud provider, as implementations should rely on the order of the addresses returned

#### Which issue(s) this PR fixes:
Fixes #

```release-note
NONE
```

/sig cloud-provider
/sig network

### CODE REVIEW

**Comment by aojea** [comment]
/assign @dims @danwinship @cheftako @thockin @andrewsykim @JoelSpeed  @elmiko 

Another one of the saga of cloud provider external and node status addresses, PTAL

File: staging/src/k8s.io/cloud-provider/node/helpers/address.go [hunk 1]
```diff
@@ -88,14 +88,12 @@ func GetNodeAddressesFromNodeIPLegacy(nodeIP net.IP, cloudNodeAddresses []v1.Nod
 	return GetNodeAddressesFromNodeIP(nodeIP.String(), cloudNodeAddresses)
 }
 
-// GetNodeAddressesFromNodeIP filters the provided list of nodeAddresses to match the
-// providedNodeIP from the Node annotation (which is assumed to be non-empty). This is
+// GetNodeAddressesFromNodeIP sorts the provided list of nodeAddresses to use the
```
Comment (danwinship) [inline staging/src/k8s.io/cloud-provider/node/helpers/address.go:91]: if you're going to change the behavior here, you should probably just sync it to the behavior of `GetNodeAddressesFromNodeIPLegacy`

**Comment by danwinship** [comment]
Description is not really clear

> There is a historic behavior that filters node addresses if the kubelet
> --node-ip flag is passed as parameters and the legacy cloud-provider was
> enabled, and this behavior was carried over to the external cloud
> provider.

The behavior of legacy cloud providers was that if you passed `--node-ip` to kubelet, then it would sort the matching IP to the front of `node.status.addresses`. The behavior of the external ccm is that if you pass `--node-ip` to kubelet, then it _removes_ all non-matching IPs from `node.status.addresses`. (I'm simplifying out non-relevant details.)

> However, this behavior can break deployments that depend on
> the cloud-controller-manager to be deployed within the cluster, as there
> is a bootstrap problem if the cloud-provider wipes out IPs.

It's not that behavior that breaks things.

The issue here is that #121028 made it so that clusters using an external cloud provider can no longer bootstrap in certain ways because kubelet is now never able to know its own node IP unless either (a) you pass it via `--node-ip` or (b) kube-apiserver and the ccm are already running. (In particular, this is breaking some OpenShift deployment models, which assumed that kubelet could deploy etcd and/or kube-apiserver via static pods and pass the node IP to them via the downward API.)

One suggested _workaround_ for the problem introduced by #121028 is for the administrator to pass `--node-ip` to kubelet, and _that_ suggestion runs into the problem that doing this would cause the node's `node.status.addresses` to be stripped of secondary addresses, which may have further side effects. This PR addresses _that_ problem.

BUT... the fact that this fixes one problem implies it could just as easily cause new ones; there could potentially be components that expect the _current_ behavior of stripping secondary addresess when you pass `--node-ip`.

Also, the entire premise of #121028 is that only the external cloud provider is able to correctly decide which IP "should" be the primary node IP, so if you say we should work around the problems it causes by using `--node-ip` then you are effectively contradicting the logic of that PR...

Of course, there isn't any perfect solution here; node IPs are a mess and need to be totally rethough.

**Comment by aojea** [comment]
> BUT... the fact that this fixes one problem implies it could just as easily cause new ones; there could potentially be components that expect the current behavior of stripping secondary addresess when you pass --node-ip.

this behavior sounds wrong to me, filtering addresses that are returned by the cloud provider based on the flag

> Also, the entire premise of https://github.com/kubernetes/kubernetes/pull/121028 is that only the external cloud provider is able to correctly decide which IP "should" be the primary node IP, so if you say we should work around the problems it causes by using --node-ip then you are effectively contradicting the logic of that PR...

well, it says the cloud-provider is authoritative, since before it wiped the others, the behavior will still be the same, no? the passed ip will be always the first one

**Comment by aojea** [comment]
> The behavior of legacy cloud providers was that if you passed `--node-ip` to kubelet, then it would sort the matching IP to the front of `node.status.addresses`. The behavior of the external ccm is that if you pass `--node-ip` to kubelet, then it _removes_ all non-matching IPs from `node.status.addresses`. (I'm simplifying out non-relevant details.)

I thought the same but checking in the old PRs I linked in the description it seems to come from the legacy ones https://github.com/kubernetes/cloud-provider/issues/56

**Comment by dims** [comment]
cc @kmala @cartermckinnon

**Comment by danwinship** [comment]
> this behavior sounds wrong to me

The relevant question (IMO) is not whether the current behavior is "right" or "wrong", it's about whether changing it will break existing clusters. We already changed the behavior to make it more "right" in #121028, and that broke things. We don't want to be back here in a few more releases to unbreak the things we break now.

Long term we certainly need better behavior but it probably needs to be opt-in. Short term I think we should either revert #121028, or add a flag to request that kubelet do the pre-#121028 behavior. (`--node-ip legacy`?)

**Comment by rphillips** [comment]
"temporary" flags are not really temporary, since we need to preserve upgrades through releases. At best, the flag's functionality would be removed, but the flag would have to stay around until a deprecation cycle and release cycle has been done.

I also believe we should revert.

**Comment by elmiko** [comment]
> "temporary" flags are not really temporary, since we need to preserve upgrades through releases.

yeah... good point, that probably pushes the needle closer to the revert option for me

**Comment by aojea** [comment]
> The relevant question (IMO) is not whether the current behavior is "right" or "wrong", it's about whether changing it will break existing clusters. We already changed the behavior to make it more "right" in #121028, and that broke things. We don't want to be back here in a few more releases to unbreak the things we break now.

the breakage here is debatable , previous behavior was a clear bug and building features on a bug 🤷 

image

what I also consider a bug is to filter addresses, there is no reason to cap information in an array of addresses where only the first ones are used by the cluster ... and the behavior with this is still the same, the addresses remain in the same position

**Comment by cartermckinnon** [comment]
On AWS, when the AWS VPC CNI is used, pod IP's are bound to the EC2 instance's ENI(s). IIUC, after this change pod IP's would show up in the node's addresses as `InternalIP`-s. We'll need to update the AWS cloud-provider to do the filtering itself

**Comment by wongma7** [comment]
> We'd need to update the AWS cloud-provider to do the filtering itself

and this (i.e. returning fewer addresses than before) would be a breaking change I would like to avoid.

> Since there is no reason to filter addresses obtained by the cloud provider, because sorting solves the problem

kubelet puts the entire list of node.status.addresses in the serving cert CSR, not just the first/preferred address of each type: https://github.com/kubernetes/kubernetes/blob/9c4181e234b7498bae16ef13c14949f3ae75c738/pkg/kubelet/certificate/kubelet.go#L92, and my CSR approver-signer is depending on the assumption that IPs get filtered such that only --node-ip(s) show up in the CSR.

> if someone is depending on only one address existing because one flag is set in the kubelet, that seems a lot

So yes it's a lot but I do i.e. my CSR approver-signer does depend on this.

I agree with the characterization of handling node.status.addresses as 'problematic' but not with --node-ip behavior being a bug. --node-ip predates node.status.addresses and I don't think any design/thought was put into its intended behavior/interaction with node.status.addresses, so reconciling --node-ip with node.status.addresses indeed happened 'holistically', and in the absence of clear documentation about its intended behavior, it's impossible to classify it as a 'bug' versus 'working-as-intended but poorly thought out and too late to change'.

**Comment by wongma7** [comment]
correction: node-ip does not predate node.status.addresses. --node-ip predates the idea that there can be multiple node.status.addresses. https://github.com/kubernetes/kubernetes/blob/a9123de9b473df7b5105708e791dad4ba3fa9992/pkg/kubelet/kubelet_node_status.go#L361

**Comment by aojea** [comment]
> after this change pod IP's would show up in the node's addresses as InternalIP

@cartermckinnon why? 

> CSR approver-signer is depending on the assumption that IPs get filtered such that only --node-ip(s) show up in the CSR.

@wongma7 I have problem to fully understand this dependency, so what happen with the CSR approver-signer if there are external IPs, do they fail? those are not filtered and the show IP , is the dependency on "there can be only one internalIP?

> , it's impossible to classify it as a 'bug' versus 'working-as-intended but poorly thought out and too late to change'.

_and not tested at all_ , anyway  it is clear we are in a bad spot  , how do we move forward then? can we do an exercise to try to get this to a better place that is easy to understand for users and developers?

**Comment by danwinship** [comment]
> i wonder if we will ultimately continue down the path of the changes that were started in https://github.com/kubernetes/kubernetes/pull/121028 ?

I've started trying to solve the node addresses problem twice (the earlier drafts of both [KEP-1664](https://github.com/kubernetes/enhancements/pull/1665) and [KEP-3705](https://github.com/kubernetes/enhancements/pull/3706)) and both times ended up chickening out and solving just a smaller piece of the problem.

But my feeling from the discussions around those KEPs is that the right long-term plan is:
- Preserve 100% backward compatibility with the existing behavior of currently-supported `--node-ip` values and `node.status.addresses`. (eg, the part of KEP-3705 that was actually implemented didn't change behavior in any previously-supported scenarios, it just defined semantics for a scenario where kubelet would previously have exited with an error.)
- Add either a new `kubelet --node-ips` flag (note plural) or else new possible values for `kubelet --node-ip` with new-and-improved semantics
  - (Renaming the flag to `--node-ips` lets us make `--node-ips 1.2.3.4` mean something different from what `--node-ip 1.2.3.4` means, if we want to improve the semantics of the "admin specifies a single IP override" case.)
- Possibly deprecate `node.status.addresses` in favor of something more well-specified (eg the `node.status.nodeIPs` from KEP-1665), or possibly just add more fields to `v1.NodeAddress` to let the cloud providers be clearer about the intended semantics of each address.

**Comment by aojea** [comment]
> Preserve 100% backward compatibility with the existing behavior of currently-supported --node-ip values and node.status.addresses. (eg, the part of KEP-3705 that was actually implemented didn't change behavior in any previously-supported scenarios, it just defined semantics for a scenario where kubelet would previously have exited with an error.)

lol, we already have this, using the unspecified address in `--node-ip 0.0.0.0` solves this problem

- it bootstraps the node without the cloud-controller-manager with the preferred IP family
- it allows the cloud controller manager to override the addresses without filtering

I tested it locally and works as we need

**Comment by aojea** [comment]
folks please comment on https://github.com/kubernetes/kubernetes/pull/125337 to document the flag, I think unspecified address save us the day again 😄 

@wongma7 I also added a test to not regress on this behavior of filtering IPs

**Comment by danwinship** [comment]
ok, filed a new issue, #125348, where we can continue discussion

**Comment by aojea** [comment]
> @wongma7 I have problem to fully understand this dependency, so what happen with the CSR approver-signer if there are external IPs, do they fail? those are not filtered and they show as node addresses , is the dependency on "there can be only one internalIP"?

ok, I was checking the code with the guidance of @liggitt ,  the filtering argument is only valid for Internal IPs, if the node has an ExternalIP or other type the CSR approver-signer will fail, no @wongma7 ?

https://github.com/kubernetes/kubernetes/blob/a298c14f18d4973a9ceaf21f1e0dc4e39b4c5bfb/pkg/kubelet/certificate/kubelet.go#L152-L187

**Comment by liggitt** [comment]
> the filtering argument is only valid for Internal IPs

what filtering argument are you referring to?

**Comment by danwinship** [comment]
>  the filtering argument is only valid for Internal IPs, if the node has an ExternalIP or other type the CSR approver-signer will fail

(btw, [notes on how different clouds assign node IPs](https://github.com/kubernetes/kubernetes/pull/123105#issuecomment-2054088415). In particular, yes, AWS does sometimes do ExternalIPs)

**Pull request closed without merging**